### ISSUE TITLE
Way to check if script loading is failed

### ISSUE THREAD

**Issue by abitrolly** [mitmproxy/mitmproxy#4438]
I spent quite a few days to debug a problem with missing import in HAR script. Now I need to write a test to fail when any exception occurs during the load, and I can find to detect if there was an exception.

For example, if `../examples/contrib/har_dump.py` is invalid, the following snippet of code will only fail if there is no `a.addons` field. Logs are empty for some reason.

```
      def test_simple(self, tmpdir, tdata):
          with taddons.context() as tctx:
              fullpath = tdata.path("../examples/contrib/har_dump.py")
              a = script.load_script(fullpath)
              assert tctx.master.logs == []
              assert a.addons
```

#### Proposal

A way to check for errors during script load. Maybe there is bug in logger. I don't know.

**Comment by mhils** [comment]
I'm not sure I understand what the problem is here. Can you give an example of what doesn't work?

FWIW, they way I develop addons is that I start `mitmproxy -s ...`, open the event log in a second pane (`-`), and then try out the functionality. If stuff doesn't work, you should be able to edit the script file and it should be automatically reloaded. Once I am happy with my addon, I write some tests that make sure it doesn't break in the future. If I understand you correctly you are working more with a test-driven workflow, is that correct?

**Comment by abitrolly** [comment]
Yep. For integration tests it is important to track API breaks. When script loading starts to fail, I expect to see the exact exception in tests.

**Comment by Prinzhorn** [comment]
Would a new flag make sense like `--scripts-mandatory` (or whatever) that will crash the mitmproxy process when an addon crashes? This might be useful for cases where you absolutely need all-or-nothing and don't want mitmproxy to continue without an addon loaded? The current architecture is great to develop addons but in production one might not want mitmproxy to "silently" continue without a given addon. Or am I misunderstanding the issue?

> I write some tests that make sure it doesn't break in the future

I think you mean to make it _less likely_ to break in the future :smile:

**Comment by abitrolly** [comment]
Not sure about the new option for `mitmproxy`. I guess it is fine if the loading error is present in the event log, but in tests the `ctx.master.logs` was empty even if script failed to parse.

So I'd expect the traceback/crash mode to be on by default for tests and `mitmdump -s <script>`. For interactive tool like `mitmproxy` it is probably not a good idea to crash.

**Comment by mhils** [comment]
We need to distinguish a bit between mitmproxy itself and our (internal) tests here.

For mitmproxy itself, I'm feeling strongly that a malformed addon should not bring down the entire proxy (that is especially true when live-reloading addons). We should print an error message in the event log instead, which I believe is happening just fine.

For tests, @abitrolly has a good point here that silently eating errors is bad. We should probably refactor [`load_script`](https://github.com/mitmproxy/mitmproxy/blob/cff98a30e20cf3d903deac5d188a26759daf1047/mitmproxy/addons/script.py#L19) to raise on error, and then catch that wherever it is called. This way tests would raise properly. Does that sound about right, or am I missing things?

**Comment by Prinzhorn** [comment]
@mhils that makes sense to me now

**Comment by abitrolly** [comment]
> For tests, @abitrolly has a good point here that silently eating errors is bad. We should probably refactor load_script to raise on error, and then catch that wherever it is called. This way tests would raise properly. Does that sound about right, or am I missing things?

In tests in particular I can see that `ctx.log`
